calyx-py/test/arb.py

```diff
@@ -0,0 +1,136 @@
+    """
+
+    wrap = prog.component("wrap")
+    wrap.input("i", 32)
+    wrap.input("j", 32)
+
+    mem1 = wrap.mem_d1("mem1", 32, 4, 32, is_ref=True)
+    mem2 = wrap.mem_d1("mem2", 32, 4, 32, is_ref=True)
+    ans = wrap.mem_d1("ans", 32, 1, 32, is_ref=True)
+
+    add_eq(wrap, "i", 0, "eq0", "i_eq_0")
+    add_eq(wrap, "i", 1, "eq1", "i_eq_1")
+    add_emit_from_mem(wrap, mem1, ans, "1")
+    add_emit_from_mem(wrap, mem2, ans, "2")
+
+    # AM, quality of life:
+    # I'd like to generate these `if` statements with the builder.
+    # I tried:
+    #   _: If = cb.if_(
+    #     port=CompPort(CompVar("eq0"), "out"),
+    #     cond=wrap.get_group("i_eq_0"),
+    #     body=Enable("emit_from_mem1"),
+    #   )
+    # but I'm running into trouble with the `port` field, which must be
+    # an ExprBuilder. On digging it looks like that's an GuardExpr in the AST,
+    # but I got stuck at that point.
+    wrap.control = ParComp(
+        [
+            If(
+                port=CompPort(CompVar("eq0"), "out"),
+                cond=CompVar("i_eq_0"),
+                true_branch=Enable("emit_from_mem1"),
+            ),
+            If(
+                port=CompPort(CompVar("eq1"), "out"),
+                cond=CompVar("i_eq_1"),
+                true_branch=Enable("emit_from_mem2"),
+            ),
+        ]
+    )
+
+
+def add_main(prog):
+    """Inserts the component `main` into the program.
+    This will be used to `invoke` the component `wrap`.
+
+    For now, I'd like to pass it memory cells `A` and `B` by reference,
+    along with the inputs i = 1, j = 3.
+    """
+    main = prog.component("main")
+    _ = main.mem_d1("A", 32, 4, 32, is_external=True)
+    _ = main.mem_d1("B", 32, 4, 32, is_external=True)
+    _ = main.mem_d1("out", 32, 1, 32, is_external=True)
+
+    # AM, point of failure (but a stupid hack has worked):
+    # I'd like to add the following to the `cells` section:
+    # together = wrap();
+
+    # The following is clearly a hilarious hack:
+    _ = main.cell("together", CompVar("wrap()"))
+    # (I've just put parentheses in the name of the component.)
+
+    # I think the following is what you actually want me to do:
+    # _ = main.cell("together", prog.component("wrap"))
+    # But _it adds a new, blank component called wrap_ to the program.
+    # I'd like for it to locate the existing component `wrap`.
+    # Thoughts?
+
+    # AM, point of failure:
+    # Maybe I'm missing something, but I think the builder library
+    # is only targeting a subset of the `invoke` functionality.
+    #   class Invoke(Control):
+    #     id: CompVar
+    #     in_connects: List[Tuple[str, Port]]
+    #     out_connects: List[Tuple[str, Port]]
+    #     ref_cells: List[Tuple[str, CompVar]] = field(default_factory=list)
+    #     comb_group: Optional[CompVar] = None
+    #     attributes: List[Tuple[str, int]] = field(default_factory=list)
+    # As I see it, only id, in_connects, and out_connects are supported.
```

Comment:
@anshumanmohan can you explain this comment? Are the other fields missing from the dataclass definition or are they unused?

Reply:
Unused! If you look at the relevant `builder` code:
https://github.com/cucapra/calyx/blob/2613b848efa3b6344edd2cd9ed3c722c889fb532/calyx-py/calyx/builder.py#L221-L239

You can see it essentially returning `ast.Invoke(id, in_connects, out_connects)`. It's leaving the other `ast.Invoke` fields blank, and then the `ast.Invoke` `doc()` method is filling in default no-ops.

Reply:
Yes, this is right—I just never hooked up the ref-cell stuff in this simple wrapper. We should probably do that somehow!